### PR TITLE
add label detection

### PR DESCRIPTION
**Pull request by jidu12**
增添标签检测功能，自测已通过

image
输出的result.json:
image

代码规范（flake8检验工具）
image

### CODE REVIEW

File: gts_engine/label_detection/label_detection.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
+import os
+import time
+import argparse
+import numpy as np
+import torch
+import json
+import random
+import sys
+sys.path.append('../')
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:9]: 不要写..，最好是os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

File: gts_engine/label_detection/label_detection.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+from sklearn.metrics import precision_score, f1_score, recall_score
+from datetime import datetime
+
+
+logger = Logger().get_log()
+
+
+def read_data(data_path, result_path):
+    input_data = []
+    with open(os.path.join(data_path, 'tnews_test.json'), 'r',
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:27]: 为啥要写死？有其他数据怎么办？

File: gts_engine/label_detection/label_detection.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+              encoding='utf8') as f:
+        for line in f:
+            input_data.append(json.loads(line.strip()))
+
+    input_labels = object
+    with open(os.path.join(data_path, 'tnews_label.json'),
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:33]: 这里也是

File: gts_engine/label_detection/label_detection.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        label_to_count[data['id']] = label_to_count.get(data['id'], 0) + 1
+        if label_to_count[data['id']] <= sample_count_every_class:
+            sample_input_data.append(data)
+    return sample_input_data
+
+
+def take_sample(sample_input_data, current_label):
+    # 样本采样  1:2 正负样本
+    total_sentences = [i['content'] for i in sample_input_data]
+    total_vail_data = sample_input_data
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:67]: vail是什么意思？

File: gts_engine/label_detection/label_detection.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    total_vail_data_index = [i for i in range(len(total_vail_data))]
+    true_label_index = [
+        i for i, k in enumerate(total_vail_data)
+        if k['label'] == current_label
+    ]
+
+    tmp_set = set(total_vail_data_index).difference(set(true_label_index))
+    negative_label_index = list(tmp_set)
+    random.shuffle(negative_label_index)
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:76]: 为什么不用random.sample

File: gts_engine/label_detection/label_detection.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    sentences = []
+    vail_data = []
+    for i in need_index:
+        sentences.append(total_sentences[i])
+        vail_data.append(total_vail_data[i])
+    return sentences, vail_data
+
+
+def label_detection(model_path, data_path):
+    start = time.time()
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:91]: 这个变量有什么作用？

File: gts_engine/label_detection/label_detection.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    starttime = datetime.now()
+    tokenizer = BertTokenizer.from_pretrained(model_path)
+    model = taskModel(model_path, tokenizer=tokenizer)
+    model.eval()
+    model.cuda()
+    random.seed(123)
+
+    result_path = os.path.join(data_path, 'label_detection_result.json')
+    input_data, input_labels, labels_info = read_data(data_path=data_path, result_path=result_path)
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:100]: 为什么要读一个result_path，不是应该写吗？

File: gts_engine/label_detection/label_detection.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    label_detection_result = []
+
+    total_count = 0
+    index = 0
+    for value in input_labels["labels"]:
+        current_label = value
+        current_label_id = index
+        index += 1
+
+        sentences, vail_data = take_sample(sample_input_data=sample_input_data, current_label=current_label)
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:111]: 这个函数，可以用sklearn的StratifiedShuffleSplit替换吗？

File: gts_engine/label_detection/label_detection.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+        total_count += len(sentences)
+        random_f1 = -1
+        label_f1 = -1
+        two_choices = [
+            '[MASK]' * len(current_label), '[MASK]' * len(current_label)
+        ]
+        n = 0
+        while n < 2:
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:120]: 为什么不把while里面的逻辑拆成函数？

File: gts_engine/label_detection/label_detection.py [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            logger.info("start {}".format(two_choices))
+            samples = []
+            for sentence in sentences:
+                tmp_sample = {
+                    "content": sentence,
+                    "label": two_choices[0]
+                }
+                samples.append(tmp_sample)
+
+            start1 = time.time()
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:130]: 这个变量有什么作用？

File: gts_engine/label_detection/label_detection.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+                score = torch.max(probs, dim=-1)[0]
+
+                probs = probs.detach().cpu().numpy()
+                predicts = predicts.detach().cpu().numpy()
+                score = score.detach().cpu().numpy()
+
+                label_idx = list(batch["label_idx"][0].numpy())
+                total_predicts += [label_idx.index(i) for i in predicts]
+
+            logger.info("every_time:{}".format(time.time() - start1))
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:172]: 这个logger不需要

File: gts_engine/label_detection/label_detection.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+            two_choices[0] = current_label
+
+        f1_info = []
+        if label_f1 > random_f1 * 1.1:
+            labels_info[current_label] = ["2", current_label_id, label_f1]
+            f1_info.append((label_f1, random_f1))
+
+        elif label_f1 <= random_f1 * 1.1 and label_f1 > random_f1:
+            if label_recall > random_recall:
+                labels_info[current_label] = ["2", current_label_id, label_f1]
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:208]: 这里labels_info应该是之前需要多次修改和检测的逻辑吧，这里还需要吗？

File: gts_engine/label_detection/label_detection.py [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+                                              endtime.minute - starttime.minute,
+                                              endtime.second - starttime.second))
+
+    with open(result_path, 'w', encoding='utf-8') as f:
+        for i in result:
+            f.write(json.dumps(i, ensure_ascii=False) + '\n')
+
+
+if __name__ == '__main__':
+    os.environ["CUDA_VISIBLE_DEVICES"] = "5"
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:257]: 环境变量由用户指定，不要写死，而且写在这里不会起作用的

File: gts_engine/label_detection/label_detection.py [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+    model_name = "Erlangshen-UniMC-MegatronBERT-1.3B-Chinese"
+    pretrained_model_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "pretrained")
+    download_model_from_huggingface(pretrained_model_dir, model_name, model_class=MegatronBertForMaskedLM, tokenizer_class=BertTokenizer)
+    # Set path to load pretrained model
+
+    total_parser = argparse.ArgumentParser()
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:263]: 先写argparse再下载模型

File: gts_engine/label_detection/label_detection.py [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+    total_parser.add_argument(
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:264]: 不需要指定model_path，因为默认都用Erlangshen-UniMC-MegatronBERT-1.3B-Chinese，这是用户不能改的

File: gts_engine/label_detection/label_detection.py [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+        "--model_path",
+        default=pretrained_model_dir,
+        type=str,
+        help="train task name")
+    total_parser.add_argument(
+        "--data_path",
+        default=os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "examples/text_classification"),
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:271]: 为啥要写死用户需要输入的数据？

**Review by pskun — CHANGES_REQUESTED** [review]

File: gts_engine/label_detection/label_detection.py (hunk 10 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:130]: 已删除

File: gts_engine/label_detection/label_detection.py (hunk 6 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:91]: 已删除

File: gts_engine/label_detection/label_detection.py (hunk 5 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:76]: 一博这里取样数量是有可能大于列表的长度的
经测试random.sample在取样数大于列表长度的情况下会报错

File: gts_engine/label_detection/label_detection.py (hunk 12 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:208]: 经测试，有一个在对原始result读取的逻辑可以删除（已完成），其他的逻辑会影响到检测过程，删除了就没法得到正确结果了

File: gts_engine/label_detection/label_detection.py (hunk 4 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:67]: 一博说这里是想表示有效数据的意思，应该是拼错了，我修改为了“val”

File: gts_engine/label_detection/label_detection.py (hunk 8 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:111]: 不太熟悉这个函数，但是我理解这个函数应该没法做到正负采样的逻辑

File: gts_engine/label_detection/label_detection.py (hunk 3 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py]: 已修改

File: README.md [hunk 17]
```diff
@@ -314,6 +314,31 @@ optional arguments:
...
+
+```json
+{"content": "请替我活下去：6岁女孩身患重病，自愿捐所有器官救5人，万人泪湿", "label": "故事", "id": 2740}
+```
+
+#### 开始检测
+一行代码启动.py文件即可开始检测
+
+```bash
+python label_detection.py --label_path /cognitive_comp/zhubohan/GTS-Engine/examples/text_classification/tnews_label.json --data_path /cognitive_comp/zhubohan/GTS-Engine/examples/text_classification/tnews_test.json
```
Comment (pskun) [inline README.md:339]: 改成相对路径

File: gts_engine/label_detection/label_detection.py [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+    with open(os.path.join(label_path),
+              'r',
+              encoding='utf8') as f:
+        input_labels = json.load(f)
+    return input_data, input_labels
+
+
+def take_sample_input(labels_length, input_data,):
+    labels_count = labels_length
+    # sample_count_every_class = int(6000/labels_count)   # 600/(0.05 * 2)
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:42]: 删掉没用的注释

File: gts_engine/label_detection/label_detection.py (hunk 5 above)
Comment (pskun) [reply gts_engine/label_detection/label_detection.py:67]: > 一博这里取样数量是有可能大于列表的长度的 经测试random.sample在取样数大于列表长度的情况下会报错

random.sample(data, min(len(data), n))

File: gts_engine/label_detection/label_detection.py (hunk 9 above)
Comment (pskun) [reply gts_engine/label_detection/label_detection.py:168]: 下一次提交代码时考虑拆分成函数

File: gts_engine/label_detection/label_detection.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+    with open(result_path, 'w', encoding='utf-8') as f:
+        for i in result:
+            f.write(json.dumps(i, ensure_ascii=False) + '\n')
+
+
+if __name__ == '__main__':
+    total_parser = argparse.ArgumentParser()
+    total_parser.add_argument(
+        "--label_path",
+        default=os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "examples/text_classification/tnews_label.json"),
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:260]: 最好不要设置default

File: gts_engine/label_detection/label_detection.py [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+        type=str,
+        help="data path")
+    total_parser.add_argument(
+        "--data_path",
+        default=os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "examples/text_classification/tnews_test.json"),
```
Comment (pskun) [inline gts_engine/label_detection/label_detection.py:265]: 这里也是

**Review by pskun — CHANGES_REQUESTED** [review]

File: README.md (hunk 17 above)
Comment (jidu12) [reply README.md:339]: > 改成相对路径

已修改 
注： 这里 刚刚测试的时候前面的.py文件如果不写全绝对路径的话前面导包（第九行）的时候会出问题，sys.path会挂一个空的路径，然后找不到父级目录的组件。。年前的时候没遇到这个问题 很奇怪

File: gts_engine/label_detection/label_detection.py (hunk 18 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:42]: > 删掉没用的注释

已修改

File: gts_engine/label_detection/label_detection.py (hunk 5 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:67]: 👌

File: gts_engine/label_detection/label_detection.py (hunk 9 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py:168]: 这里while拆过一个 total_predicts = predict_process(two_choices=two_choices,
                                             tokenizer=tokenizer,
                                             sentences=sentences,
                                             model=model)
函数，其他的部分耦合度太高，拆的话要传的参数太多了

File: README.md (hunk 17 above)
Comment (jidu12) [reply README.md:339]: 已修改

File: gts_engine/label_detection/label_detection.py (hunk 19 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py]: 好滴

File: gts_engine/label_detection/label_detection.py (hunk 20 above)
Comment (jidu12) [reply gts_engine/label_detection/label_detection.py]: 好的

**Review by pskun — APPROVED** [review]
